### PR TITLE
re-use database connections between tests

### PR DESCRIPTION
**Pull request by ewollesen**
This results in massive speedups for the effected package (data/store/mongo). Compare for
`go clean -cache ; make ci-test TEST=data/store/mongo`:

    Ginkgo ran 1 suite in 4m51.037201582s

  -- vs --

    Ginkgo ran 1 suite in 38.723898055s

...for a speed up of 7.5x.

And for `go clean -cache ; make test TEST=data/store/mongo`:

    Ginkgo ran 1 suite in 4m41.735427524s

  -- vs --

    Ginkgo ran 1 suite in 24.292654343s

...this time the speedup is 11.5x.

The best part? This technique can be used in other tests within our code base for more gains.

### CODE REVIEW

File: data/store/mongo/mongo_test.go [hunk 1]
```diff
@@ -237,35 +237,42 @@ func DataSetDatumAsInterface(dataSetDatum data.Datum) interface{} {
...
+				Expect(err).To(Succeed())
+			})
+		})
 	})
 
 	Context("New", func() {
 		It("returns an error if unsuccessful", func() {
-			var err error
-			store, err = dataStoreMongo.NewStore(nil)
+			store, err := dataStoreMongo.NewStore(nil)
```
Comment (darinkrauss) [inline data/store/mongo/mongo_test.go:269]: [nit] This is a little confusion with `store` matching the same name as the more "global" `store`. Consider using `testStore` as you did above. Same for the test below.

Comment (darinkrauss) [reply data/store/mongo/mongo_test.go:269]: Or, you could pull these two tests out of the BeforeEach/AfterEach Describe. (You'd need to create a new Context with for most everything and put these outside of that.)

File: data/store/mongo/mongo_test.go [hunk 2]
```diff
@@ -276,26 +283,29 @@ var _ = Describe("Mongo", Label("mongodb", "slow", "integration"), func() {
...
-				collection.Database().Drop(context.Background())
-				dataSetCollection.Database().Drop(context.Background())
-				summaryCollection.Database().Drop(context.Background())
-				alertsCollection.Database().Drop(context.Background())
-			}
+			var err error
+			ctx := context.Background()
+			all := bson.D{}
+			_, err = collection.DeleteMany(ctx, all)
+			Expect(err).To(Succeed())
```
Comment (darinkrauss) [inline data/store/mongo/mongo_test.go:302]: [nit] I think you can rewrite this as:
```suggestion
            Expect(collection.DeleteMany(ctx, all)).Error().ShouldNot(HaveOccurred())
```

Comment (darinkrauss) [reply data/store/mongo/mongo_test.go:302]: And the ones below.

File: store/structured/mongo/test/test.go [hunk 3]
```diff
@@ -18,6 +19,9 @@ import (
 var database string
 var address string
 
+var cleanupHandlers []func()
+var cleanupHandlersMu sync.Mutex
```
Comment (darinkrauss) [inline store/structured/mongo/test/test.go:23]: [nit] `cleanupHandlersMutex` so it is clear?

File: store/structured/mongo/test/test.go [hunk 4]
```diff
@@ -30,8 +34,32 @@ var _ = ginkgo.AfterSuite(func() {
...
+// DeferredCleanup provides a way to have something run during AfterSuite.
+//
+// Ginkgo only allows a single AfterSuite to be instantiated, and it's here in this file, so
+// give other tests a chance to make use of it.
+func DeferredCleanup(f func()) {
+	cleanupHandlersMu.Lock()
+	defer cleanupHandlersMu.Unlock()
+	if cleanupHandlers == nil {
+		cleanupHandlers = []func(){}
+	}
```
Comment (darinkrauss) [inline store/structured/mongo/test/test.go:49]: I believe `append` will do this for you?

**Review by darinkrauss — CHANGES_REQUESTED** [review]
One change to remove unnecessary code, plus a few nits for your consideration.

File: data/store/mongo/mongo_test.go (hunk 1 above)
Comment (ewollesen) [reply data/store/mongo/mongo_test.go:269]: Will do.

File: store/structured/mongo/test/test.go (hunk 3 above)
Comment (ewollesen) [reply store/structured/mongo/test/test.go:23]: I believe the -Mu suffix is well established as indicating a sync.Mutex.

If I grep for `'Mu\s+sync.Mutex'` I get 41 hits.

If I grep for `'Mutex\s+sync.Mutex'` I get 3 hits.

The numbers are similar when searching the Go stdlib code (40 and 8).

File: store/structured/mongo/test/test.go (hunk 4 above)
Comment (ewollesen) [reply store/structured/mongo/test/test.go:49]: > I believe `append` will do this for you?

True, but this code was removed later, so it doesn't matter this time.

File: data/store/mongo/mongo_test.go (hunk 1 above)
Comment (ewollesen) [reply data/store/mongo/mongo_test.go:269]: To be clear, this change will appear in #777 not here.

**Comment by ewollesen** [comment]
All comments addressed in #777

**Pull request closed without merging**